app.ts

```diff
@@ -0,0 +1,1160 @@
+					res = await gw.zwave.storeDevices(
+						data.devices,
+						data.nodeId,
+						data.remove
+					)
+					break
+			}
+		} catch (error) {
+			logger.error('Error while calling HASS api', error)
+			err = error.message
+		}
+
+		const result = {
+			success: !err,
+			message: err || 'Success HASS api call',
+			result: res,
+			api: data.apiName,
+		}
+
+		socket.emit(socketEvents.api, result)
+	})
+}
+
+// ### APIs
+
+function isAuthEnabled() {
+	const settings = jsonStore.get(store.settings)
+	return settings.gateway && settings.gateway.authEnabled === true
+}
+
+async function parseJWT(req: Request) {
+	// if not authenticated check if he has a valid token
+	let token = req.headers['x-access-token'] || req.headers.authorization // Express headers are auto converted to lowercase
+	token = Array.isArray(token) ? token[0] : token
+	if (token && token.startsWith('Bearer ')) {
+		// Remove Bearer from string
+		token = token.slice(7, token.length)
+	}
+
+	// third-party cookies must be allowed in order to work
+	if (!token) {
+		throw Error('Invalid token header')
+	}
+	const decoded = await verifyJWT(token, sessionSecret)
+
+	// Successfully authenticated, token is valid and the user _id of its content
+	// is the same of the current session
+	const users = jsonStore.get(store.users)
+
+	const user = users.find(
+		(u: { username: any }) => u.username === decoded.username
+	)
+
+	if (user) {
+		return user
+	} else {
+		throw Error('User not found')
+	}
+}
+
+// middleware to check if user is authenticated
+async function isAuthenticated(req: Request, res: Response, next: () => void) {
+	// if user is authenticated in the session, carry on
+	if (req.session.user || !isAuthEnabled()) {
+		return next()
+	}
+
+	// third-party cookies must be allowed in order to work
+	try {
+		const user = await parseJWT(req)
+		req.user = user
+		next()
+	} catch (error) {
+		logger.debug('Authentication failed', error)
+	}
+
+	res.json({
+		success: false,
+		message: RESPONSE_CODES['3'],
+		code: 3,
+	})
+}
+
+// logout the user
+app.get('/api/auth-enabled', apisLimiter, function (req, res) {
+	res.json({ success: true, data: isAuthEnabled() })
+})
+
+// api to authenticate user
+app.post(
+	'/api/authenticate',
+	loginLimiter,
+	csrfProtection,
+	async function (req, res) {
+		const token = req.body.token
+		let user: User
+
+		try {
+			// token auth, mostly used to restore sessions when user refresh the page
+			if (token) {
+				const decoded = await verifyJWT(token, sessionSecret)
+
+				// Successfully authenticated, token is valid and the user _id of its content
+				// is the same of the current session
+				const users = jsonStore.get(store.users)
+
+				user = users.find((u: User) => u.username === decoded.username)
+			} else {
+				// credentials auth
+				const users = jsonStore.get(store.users)
+
+				const username = req.body.username
+				const password = req.body.password
+
+				user = users.find((u: User) => u.username === username)
```

Comment:
try to avoid inline function type declarations

Reply:
dopne